In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import os

In [8]:
path = 'data/IND'

def list_all_files_and_folders(directory, files = []):
    for entry in os.listdir(directory):
        full_path = os.path.join(directory, entry)
        if os.path.isdir(full_path):
            list_all_files_and_folders(full_path, files)  # Recursive call for subdirectories
        else:
            files.append(full_path[8:])
    return files

files = list_all_files_and_folders(path)

In [9]:
files

['\\3MINDIA.NS.csv',
 '\\AARTIIND.NS.csv',
 '\\ABAN.NS.csv',
 '\\ABB.NS.csv',
 '\\ACC.NS.csv',
 '\\ADANIENT.NS.csv',
 '\\ADORWELD.NS.csv',
 '\\AEGISCHEM.NS.csv',
 '\\AFTEK.NS.csv',
 '\\AGRODUTCH.NS.csv',
 '\\AKZOINDIA.NS.csv',
 '\\ALPSINDUS.NS.csv',
 '\\AMBUJACEM.NS.csv',
 '\\APOLLOHOSP.NS.csv',
 '\\APTECHT.NS.csv',
 '\\ARCHIES.NS.csv',
 '\\ARVIND.NS.csv',
 '\\ASAL.NS.csv',
 '\\ASHIMASYN.NS.csv',
 '\\ASHOKLEY.NS.csv',
 '\\ASIL.NS.csv',
 '\\ASTRAZEN.NS.csv',
 '\\ATFL.NS.csv',
 '\\ATNINTER.NS.csv',
 '\\ATUL.NS.csv',
 '\\AUSOMENT.NS.csv',
 '\\AUTOAXLES.NS.csv',
 '\\AXISBANK.NS.csv',
 '\\BAGFILMS.NS.csv',
 '\\BAJAJ-AUTO.NS.csv',
 '\\BAJAJFINSV.NS.csv',
 '\\BAJAJHIND.NS.csv',
 '\\BAJFINANCE.NS.csv',
 '\\BANARISUG.NS.csv',
 '\\BANKINDIA.NS.csv',
 '\\BASF.NS.csv',
 '\\BATAINDIA.NS.csv',
 '\\BBTC.NS.csv',
 '\\BEL.NS.csv',
 '\\BEML.NS.csv',
 '\\BHARATRAS.NS.csv',
 '\\BHARTIARTL.NS.csv',
 '\\BHEL.NS.csv',
 '\\BIL.NS.csv',
 '\\BINANIIND.NS.csv',
 '\\BLUECHIP.NS.csv',
 '\\BLUEDART.NS.csv',
 '\\BOM

In [30]:
data = pd.read_csv("data/can/ABX.csv")

In [7]:
x_close = data["x_close"].to_numpy()

In [19]:
# Initialize arrays for skewness and kurtosis
skewness_values = np.zeros_like(x_close)
kurtosis_values = np.zeros_like(x_close)

# Calculate skewness and kurtosis in groups of 10
for i in range(0, len(x_close)-10):
    group = x_close[i:i+10]
    if len(group) > 1:  # Ensure at least 2 elements for valid calculation
        skewness_values[i+10] = stats.skew(group, bias=False)  # Unbiased estimator
        kurtosis_values[i+10] = stats.kurtosis(group, bias=False)  # Unbiased estimator

print(x_close)
print(skewness_values)
print(kurtosis_values)

[ 0.          0.01378273 -0.0124251  ...  0.00415675  0.01393443
  0.02112328]
[0.         0.         0.         ... 0.50848534 0.26160311 0.30123057]
[ 0.          0.          0.         ... -0.0989411   0.38295351
  0.70709124]


In [25]:
# Insert the new columns at the desired position (between x_close and y_close)
data.insert(loc=2, column="Skewness", value=skewness_values)  # Insert 'None' values initially
data.insert(loc=3, column="Kurtosis", value=kurtosis_values)  # Insert 'None' values initially

# You can now assign desired values to the new columns

# Print the resulting DataFrame
print(data)

            date   x_close  Skewness  Kurtosis   y_close
0     2003-01-02  0.000000  0.000000  0.000000  0.987652
1     2003-01-03  0.013783  0.000000  0.000000  0.976159
2     2003-01-06 -0.012425  0.000000  0.000000  1.020959
3     2003-01-07 -0.024130  0.000000  0.000000  0.981850
4     2003-01-08  0.020742  0.000000  0.000000  1.014627
5     2003-01-09 -0.018317  0.000000  0.000000  0.991904
6     2003-01-10  0.014521  0.000000  0.000000  0.969808
7     2003-01-13 -0.008129  0.000000  0.000000  1.021330
8     2003-01-14 -0.030657  0.000000  0.000000  1.035547
9     2003-01-15  0.021106  0.000000  0.000000  0.990362
10    2003-01-16  0.034930 -0.080347 -1.612055  1.003716
11    2003-01-17 -0.009685 -0.005938 -1.550725  0.990578
12    2003-01-20  0.003709  0.349287 -1.337667  1.025472
13    2003-01-21 -0.009467  0.108761 -1.286484  1.037898
14    2003-01-22  0.025153  0.084549 -0.912932  1.026992
15    2003-01-23  0.037198  0.087857 -1.053043  0.958955
16    2003-01-24  0.026634 -0.2

In [35]:
for file in files:
    data = pd.read_csv("data/can" + file)
    x_close = data["x_close"].to_numpy()
    
    # Initialize arrays for skewness and kurtosis
    skewness_values = np.zeros_like(x_close)
    kurtosis_values = np.zeros_like(x_close)
    
    # Calculate skewness and kurtosis in groups of 10
    for i in range(0, len(x_close)-10):
        group = x_close[i:i+10]
        if len(group) > 1:  # Ensure at least 2 elements for valid calculation
            skewness_values[i+10] = stats.skew(group, bias=False)  # Unbiased estimator
            kurtosis_values[i+10] = stats.kurtosis(group, bias=False)  # Unbiased estimator
            
    # Insert the new columns at the desired position (between x_close and y_close)
    data.insert(loc=2, column="Skewness", value=skewness_values)  # Insert 'None' values initially
    data.insert(loc=3, column="Kurtosis", value=kurtosis_values)  # Insert 'None' values initially  
    
    data.to_csv("data/can_new" + file, index=False)

Canada New Data
Results for category: can_new_waveCorr
--------------------------------------------------
Annual return   :27.0\% (0.0\%)
Annual vol      :14.0\% (0.0\%)
SR     :1.88 (0.0)
MDD    :15.0\% (0.0\%)
Daily hit rate  :53.0\% (0.0\%)
Turnover        :0.14 (0.0)
GE        :0.97 (0.0)
Annual return EW:4.0\% (0\%)
Annual vol EW   :14.0\% (0\%)
SR EW  :0.28 (0)
MDD EW :36.0\% (0\%)
--------------------------------------------------
Results for category: can_new_cs_LSTM_CNN
--------------------------------------------------
Annual return   :14.0\% (0.0\%)
Annual vol      :22.0\% (0.0\%)
SR     :0.64 (0.0)
MDD    :34.0\% (0.0\%)
Daily hit rate  :52.0\% (0.0\%)
Turnover        :0.25 (0.0)
GE        :1.0 (0.0)
Annual return EW:4.0\% (0\%)
Annual vol EW   :14.0\% (0\%)
SR EW  :0.28 (0)
MDD EW :36.0\% (0\%)
--------------------------------------------------
Results for category: can_new_cs_CNN
--------------------------------------------------
Annual return   :14.0\% (0.0\%)
Annual vol      :21.0\% (0.0\%)
SR     :0.68 (0.0)
MDD    :36.0\% (0.0\%)
Daily hit rate  :52.0\% (0.0\%)
Turnover        :0.24 (0.0)
GE        :1.0 (0.0)
Annual return EW:4.0\% (0\%)
Annual vol EW   :14.0\% (0\%)
SR EW  :0.28 (0)
MDD EW :36.0\% (0\%)
--------------------------------------------------
Results for category: can_new_eiie
--------------------------------------------------
Annual return   :3.0\% (0.0\%)
Annual vol      :15.0\% (0.0\%)
SR     :0.17 (0.0)
MDD    :37.0\% (0.0\%)
Daily hit rate  :49.0\% (0.0\%)
Turnover        :0.04 (0.0)
GE        :0.87 (0.0)
Annual return EW:4.0\% (0\%)
Annual vol EW   :14.0\% (0\%)
SR EW  :0.28 (0)
MDD EW :36.0\% (0\%)
--------------------------------------------------